subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java

```diff
@@ -277,6 +231,96 @@ private void writeFailureDetails(StyledTextOutput output, FailureDetails details
+            return failure;
+        }
+
+        private FailureDetails getFailureDetails(String granularity, ExceptionStyle exceptionStyle) {
+            FailureDetails details = new FailureDetails(failure, exceptionStyle);
+            details.summary.format("%s failed with an exception.", granularity);
+
+            fillInFailureResolution(details);
+
+            if (failure instanceof ContextAwareException) {
+                ((ContextAwareException) failure).accept(new ExceptionFormattingVisitor(details));
+            } else {
+                details.appendDetails();
+            }
+            details.renderStackTrace();
+            return details;
+        }
+
+        private void fillInFailureResolution(FailureDetails details) {
+            BufferingStyledTextOutput resolution = details.resolution;
+            ContextImpl context = new ContextImpl(resolution);
+            if (details.failure instanceof FailureResolutionAware) {
+                ((FailureResolutionAware) details.failure).appendResolutions(context);
+            }
+            if (details.exceptionStyle == ExceptionStyle.NONE) {
+                context.appendResolution(output -> {
+                    resolution.text("Run with ");
+                    resolution.withStyle(UserInput).format("--%s", LoggingConfigurationBuildOptions.StacktraceOption.STACKTRACE_LONG_OPTION);
+                    resolution.text(" option to get the stack trace.");
+                });
+            }
+            if (loggingConfiguration.getLogLevel() != LogLevel.DEBUG) {
+                context.appendResolution(output -> {
+                    resolution.text("Run with ");
+                    if (loggingConfiguration.getLogLevel() != LogLevel.INFO) {
+                        resolution.withStyle(UserInput).format("--%s", LoggingConfigurationBuildOptions.LogLevelOption.INFO_LONG_OPTION);
+                        resolution.text(" or ");
+                    }
+                    resolution.withStyle(UserInput).format("--%s", LoggingConfigurationBuildOptions.LogLevelOption.DEBUG_LONG_OPTION);
+                    resolution.text(" option to get more log output.");
+                });
+            }
+
+            if (!context.missingBuild && !gradleEnterprisePluginApplied) {
+                addBuildScanMessage(context);
+            }
+        }
+
+        private void addBuildScanMessage(ContextImpl context) {
+            context.appendResolution(output -> {
+                output.text("Run with ");
+                output.withStyle(UserInput).format("--%s", StartParameterBuildOptions.BuildScanOption.LONG_OPTION);
+                output.text(" to get full insights.");
+            });
+        }
+    }
+
+    private static boolean isGradleEnterprisePluginApplied(BuildResult result) {
+        GradleInternal gradle = (GradleInternal) result.getGradle();
+        return gradle != null && gradle.getServices().get(GradleEnterprisePluginManager.class).isPresent();
```

Comment:
Can `gradle.getServices()` or `gradle.getServices().get(GradleEnterprisePluginManager.class)` ever return null?

Reply:
Instead of getting the service from the build result, could this be injected into the BuildExceptionReporter?

Reply:
I think it's pretty unlikely that `getServices()` would return null.  Or rather, if it did, a bunch of other stuff would go up in flames along with this.  And I don't think we construct a `Gradle` object anywhere that isn't backed by `BuildTreeScopeServices`, but that's maybe a more plausible failure case - that a `GradleEnterprisePluginManager` is not in the service registry.  It wouldn't be null, of course, but `get()` would return an `UnknownServiceException`.

I've changed it to handle such a case gracefully as we'd rather have the build failure than some weird service exception.

Reply:
Sorry @big-guy I didn't see your message until after I had already posted my previous reply (wasn't ignoring you).  I think we definitely want to inject this in some way, but `BuildExceptionReporter` is constructed way early, before any service registries or anything are created and all of the context that we care about comes along later and is specific to an individual build itself (just like the build result).  I think we could probably extract some sort of `BuildFailureRenderer` or something like that which is build-specific and travels with the `BuildResult` (and gets injected with what it needs).  Or we could populate the failure details and pass something along with the build result more rich than just a throwable.  It seems like there are some ways to refactor all fo this stuff, but I don't see a clear path just yet.

Reply:
Actually, now that I'm digging through all the layers of this thing, I'm understanding better how this works and grokking @ldaley's comment a lot better.  I was _really_ confused by the name `ToolingBuildTreeScopeServices` (and its package) thinking this only applied to tooling api invocations and didn't realize this was being used in command line invocations as well.  I think I see where to go with this now.

Reply:
Man, I completely misunderstood how `BuildExceptionReporter` was being invoked in command line builds and thought it would require a much more significant refactor to get `GradleEnterprisePluginManager` injected in some way.  

_shakes head in disgust_  

In any event, I've reworked this so that this is all getting injected via service infrastructure along the lines of what @ldaley suggested.  Much better now.